### PR TITLE
Redesigned pairing flow and code refactor

### PR DESCRIPTION
**Pull request by chrisveilleux**
#### Description
In preparation for the Mark II rollout, a minor rework was done of the pairing sequence.  This opportunity was also taken to perform a major refactoring of the code.

The changes to the pairing sequence include 
* a new screen asking the user to create an account before pairing the device.
* a new screen indicating pairing success
* new URL for pairing

The refactoring includes 
* reorganizing the code
* renaming some ambiguously named attributes
* making most methods private
* adding docstrings where they were lacking
* abstracting duplicate code

#### Type of PR
- [ ] Bugfix
- [X] Feature implementation
- [X] Refactor of code (without functional changes)
- [ ] Documentation improvements
- [ ] Test improvements

#### Testing
Remove the identity file from your device and reboot it.  This should initiate the pairing sequence

#### Documentation
Docstrings were added where they were lacking.  Internal specification document was also updated.

### CODE REVIEW

**Comment by forslund** [comment]
Looks really nice, gonna look at it in more detail later. You seem to have committed a bunch of IDE related files along with the changes

File: __init__.py [hunk 1]
```diff
@@ -12,52 +12,71 @@
...
+        self.pairing_code_expiration = None
         self.state = str(uuid4())
-        self.activator = None
-        self.activator_lock = Lock()
-        self.activator_cancelled = False
-
+        self.platform = None
+        self.nato_alphabet = None
         self.counter_lock = Lock()
-        self.count = -1  # for repeating pairing code. -1 = not running
```
Comment (forslund) [inline __init__.py:47]: the self.counter_lock() can be removed since it's not used anymore.

File: __init__.py [hunk 2]
```diff
@@ -66,163 +85,261 @@ def initialize(self):
...
-                    # We snuck in to this handler somehow while the pairing
-                    # process is still being setup.  Ignore it.
-                    self.log.debug("Ignoring call to handle_pairing")
-                    return
-                # Not paired or already pairing, so start the process.
-                self.count = 0
-            self.reload_skill = False  # Prevent restart during the process
-
-            self.log.debug("Kicking off pairing sequence")
+            self.log.info("Pairing skill invoked but device is paired, exiting")
```
Comment (forslund) [inline __init__.py]: Could possibly be debug level

Also pycodestyle marks this as beeing too long 80 > 79

File: __init__.py [hunk 3]
```diff
@@ -66,163 +85,261 @@ def initialize(self):
...
-
-            if not self.activator:
-                self.__create_activator()
-
-    def check_for_activate(self):
-        """Method is called every 10 seconds by Timer. Checks if user has
-        activated the device yet on home.mycroft.ai and if not repeats
-        the pairing code every 60 seconds.
+        else:
+            self.enclosure.mouth_text("mycroft.ai/pair      ")
```
Comment (forslund) [inline __init__.py:209]: This url gives me a 404 not found

File: __init__.py [hunk 4]
```diff
@@ -12,52 +12,71 @@
...
+        self.pairing_status_lock = Lock()
+        self.pairing_in_progress = False
 
     def initialize(self):
+        """Stuff to do after constructor but before intent."""
         self.add_event("mycroft.not.paired", self.not_paired)
-        self.nato_dict = self.translate_namedvalues('codes')
+        self.nato_alphabet = self.translate_namedvalues('codes')
+        # TODO replace self.platform logic with call to enclosure capabilities
+        self.platform = self.config_core['enclosure'].get('platform', 'unknown')
```
Comment (forslund) [inline __init__.py]: pycodestyle marks this as beeing too long 80 > 79

File: __init__.py [hunk 5]
```diff
@@ -12,52 +12,71 @@
...
+        self.paired_dialog = None
         self.pairing_performed = False
 
-        self.num_failed_codes = 0
+        # These attributes are used when determining if pairing has started.
+        self.pairing_status_lock = Lock()
+        self.pairing_in_progress = False
 
     def initialize(self):
+        """Stuff to do after constructor but before intent."""
```
Comment (forslund) [inline __init__.py]: Maybe write something like

```
"""
Setup language and platform dependent info and register any needed
event handlers.
"""

**Review by forslund — APPROVED** [review]
Very nice, quite clear and the important logic seems to be the same.

I noticed the new /pair url doesn't work, so this should probably not be merged before that is published on the backend.

I don't see any real need of making the methods private (`_xxx()`), since this is not loaded like a normal python module and only the skill itself will have access to the members. It only makes the code harder to read.

I Marked some things noted by linting and one or two minor edits which are totally optional.

I haven't tried this on a QT gui yet so I might have missed something there.

File: __init__.py (hunk 2 above)
Comment (chrisveilleux) [reply __init__.py]: fixed

File: __init__.py (hunk 4 above)
Comment (chrisveilleux) [reply __init__.py]: fixed

File: __init__.py (hunk 1 above)
Comment (chrisveilleux) [reply __init__.py:47]: good catch.  removed.

File: __init__.py (hunk 3 above)
Comment (chrisveilleux) [reply __init__.py:209]: agree that this PR should not merge until the URL is activated.

**Comment by chrisveilleux** [comment]
Thanks for looking at this, Åke.  IDE files removed.  I also addressed your other concerns.  One exception being private methods.  I don't think making methods private should be situational.  Consistency is key.  Public methods define the API into the code.  Even if it is implied.

**Comment by forslund** [comment]
Totally valid point, I don't quite agree but it's such a tiny thing so it's fine either way. Possibly my exposure to c++ and rust that has made me hate extra special characters that aren't really needed :)

File: __init__.py (hunk 3 above)
Comment (krisgesling) [reply __init__.py:209]: Added redirect

**Comment by krisgesling** [comment]
More changes coming but this looks good to push.

**Pull request merged**